README


This archive contains the distribution Catalyst-Model-REST,
version 0.26:

  REST model class for Catalyst

This software is copyright (c) 2012 by Kaare Rasmussen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Catalyst-Model-REST

0.26    2012-02-20 17:48:25 Europe/Copenhagen

0.25    2012-02-17 20:47:06 Europe/Copenhagen
  - JSON data can be a hashref or an arrayref (Thanks to Alex Howarth for reporting)

0.24    2011-10-10 19:23:36 Europe/Copenhagen
  - Better handling of the return content. Perhaps the content isn't
    deserializable

0.23    2011-10-09 16:19:03 Europe/Copenhagen

0.22    2011-07-29 06:06:53 Europe/Copenhagen
  - Fix error in GET call (Warming)

0.21    2011-07-20 10:23:23 Europe/Copenhagen
  - Changed Test requirement from JSON::XS to JSON
  - Make serializer test more accurate

0.20    2011-06-25 18:13:10 Europe/Copenhagen
  - New args parameter for all REST calls

0.19    2011-06-22 17:43:34 Europe/Copenhagen
  - Fix calling errors and remove charset from content_type (warming)
  - Catch (de)serialization errors

0.18    2011-06-09 17:27:34 Europe/Copenhagen
  - Uri escape parameters for all GET calls
  - Fix test dependencies

0.17    2011-06-07 14:33:13 Europe/Copenhagen
  - Fix calling parameters for calling w/o content data
  - Enhance documentation
  - Remove dependency for Data::Serializable. Makes CMR code cleaner.

0.16    2011-04-14 21:42:13 Europe/Copenhagen
 - Add clientattr to be able to set HTTP::Tiny attributes
 - Special casing application/x-www-form-urlencoded. Idea and test: Giovanni Pensa

0.15    2011-03-29 21:20:27 Europe/Copenhagen
 - Return an error if status 5XX

0.14    2011-01-27 12:28:50 Europe/Copenhagen
 - Replace LWP with HTTP::Tiny
 - Internal refactoring of methods
 - New options method

0.13    2010-10-31 08:28:27 Europe/Copenhagen
 - Force dzil to require the correct version of Data::Serialized

0.12    2010-10-30 11:46:47 Europe/Copenhagen
 - Make dzil do more
 - Added test requirements to satisfy serialization tests

0.11    26 October 2010
 - Let dzil handle version numbers in source files
 - Add JSON::XS to test requirements in dist.ini

0.10    23 October 2010
 - Use Dist::Zilla

0.09    11 October 2010
 - Allow for no data to be sent
 - Include all methods, including DELETE

0.08    7 October 2010
 - *NB* The attribute "type" now uses the content-type, so valid values are
   application/json, application/xml, application/yaml.
   More may come if needed.

0.07    5 October 2010
 - Data::Serializable needs to be 0.40.1 for XML tests to pass

0.06    3 October 2010
 - Implements serializers through Data::Serializable. This makes it much easier
   to add a new serializer if needed.

0.05    30 September 2010
 - Return a response object instead of a hashref from the REST methods

0.04    29 September 2010
 - Only deserialize if OK response

0.03    29 September 2010
 - Added code, response ro atributes

0.02    26 September 2010
 - Serializers, now two; json and yaml. Implemented as roles and controlled w/an enum
   Makefile.PL rearranged

0.01    23 September 2010
 - First version, released on an unsuspecting world.


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Data::Serializer" => 0,
    "Data::Serializer::JSON" => 0,
    "Data::Serializer::XML::Simple" => 0,
    "Data::Serializer::YAML" => 0,
    "JSON" => "2.00",
    "Module::Build" => "0.3601",
    "Test::More" => "0.88",
    "XML::Simple" => 0,
    "YAML" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "REST model class for Catalyst",
  "dist_author" => [
    "Kaare Rasmussen <kaare at cpan dot net>"
  ],
  "dist_name" => "Catalyst-Model-REST",
  "dist_version" => "0.26",
  "license" => "perl",
  "module_name" => "Catalyst::Model::REST",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Catalyst::Model" => 0,
    "Moose" => 0,
    "Role::REST::Client" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "REST model class for Catalyst",
  "AUTHOR" => "Kaare Rasmussen <kaare at cpan dot net>",
  "BUILD_REQUIRES" => {
    "Data::Serializer" => 0,
    "Data::Serializer::JSON" => 0,
    "Data::Serializer::XML::Simple" => 0,
    "Data::Serializer::YAML" => 0,
    "JSON" => "2.00",
    "Module::Build" => "0.3601",
    "Test::More" => "0.88",
    "XML::Simple" => 0,
    "YAML" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Catalyst-Model-REST",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Catalyst::Model::REST",
  "PREREQ_PM" => {
    "Catalyst::Model" => 0,
    "Moose" => 0,
    "Role::REST::Client" => 0
  },
  "VERSION" => "0.26",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Catalyst/Model/REST.pm
package Catalyst::Model::REST;
{
  $Catalyst::Model::REST::VERSION = '0.26';
}

use Moose;

extends 'Catalyst::Model';

with 'Role::REST::Client';


__PACKAGE__->meta->make_immutable;

1;


=pod

=head1 NAME

Catalyst::Model::REST - REST model class for Catalyst

=head1 VERSION

version 0.26

=head1 SYNOPSIS

Use from a controller

	# model
	__PACKAGE__->config(
		server =>      'http://localhost:3000',
		type   =>      'application/json',
		clientattrs => {timeout => 5},
	);

	# controller
	sub foo : Local {
		my ($self, $c) = @_;
		my $res = $c->model('MyData')->post('foo/bar/baz', {foo => 'bar'});
		my $code = $res->code;
		my $data = $res->data;
		...
	}

For internal use

       # model
       sub model_foo {
               my ($self) = @_;
               my $res = $self->post('foo/bar/baz', {foo => 'bar'});
               my $code = $res->code;
               my $data = $res->data;
               return $data if $code == 200;
       }

=head1 DESCRIPTION

This Catalyst Model class is a thin wrapper over L<Role::Model::Client>

Use this module if you need to talk to a REST server as a separate model.

=head1 NAME

Catalyst::Model::REST - REST model class for Catalyst

=head1 AUTHOR

Kaare Rasmussen, <kaare at cpan dot com>

=head1 BUGS 

Please report any bugs or feature requests to bug-catalyst-model-rest at rt.cpan.org, or through the
web interface at http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Catalyst-Model-REST.

=head1 COPYRIGHT & LICENSE 

Copyright 2012 Kaare Rasmussen, all rights reserved.

This library is free software; you can redistribute it and/or modify it under the same terms as 
Perl itself, either Perl version 5.8.8 or, at your option, any later version of Perl 5 you may 
have available.

=head1 AUTHOR

Kaare Rasmussen <kaare at cpan dot net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Kaare Rasmussen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__
# ABSTRACT: REST model class for Catalyst


lib/Catalyst/Model/REST/Response.pm
package Catalyst::Model::REST::Response;
{
  $Catalyst::Model::REST::Response::VERSION = '0.25';
}
use 5.010;
use Moose;

has 'code' => (
    isa => 'Int',
    is  => 'ro',
);
has 'response' => (
    isa => 'HashRef | ArrayRef',
    is  => 'ro',
);
has 'error' => (
    isa => 'Str',
    is  => 'ro',
);
has 'data' => (
    isa => 'HashRef',
    is  => 'ro',
);

__PACKAGE__->meta->make_immutable;

1;


=pod

=head1 NAME

Catalyst::Model::REST::Response

=head1 VERSION

version 0.25

=head1 NAME

Catalyst::Model::REST::Response - Response class for REST

=head1 METHODS

=head2 code

Returns the http status code of the request

=head2 response

Returns the raw HTTP::Tiny response. Use this if you need more information than status and content.

=head2 error

Returns the returned reason from HTTP::Tiny where the status is 500 or higher. 

=head2 data

Returns the deserialized data. Returns an empty hashref if the response was unsuccessful

=head1 AUTHOR

Kaare Rasmussen, <kaare at cpan dot com>

=head1 BUGS 

Please report any bugs or feature requests to bug-catalyst-model-rest at rt.cpan.org, or through the
web interface at http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Catalyst-Model-REST.

=head1 COPYRIGHT & LICENSE 

Copyright 2010 Kaare Rasmussen, all rights reserved.

This library is free software; you can redistribute it and/or modify it under the same terms as 
Perl itself, either Perl version 5.8.8 or, at your option, any later version of Perl 5 you may 
have available.

=head1 AUTHOR

Kaare Rasmussen <kaare at cpan dot net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Kaare Rasmussen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__


lib/Catalyst/Model/REST/Serializer.pm
package Catalyst::Model::REST::Serializer;
{
  $Catalyst::Model::REST::Serializer::VERSION = '0.25';
}
use 5.010;
use Try::Tiny;
use Moose;
use Moose::Util::TypeConstraints;

use Data::Serializer;

has 'type' => (
    isa => enum ([qw{application/json application/xml application/yaml application/x-www-form-urlencoded}]),
    is  => 'rw',
	default => 'application/json',
);
no Moose::Util::TypeConstraints;
has 'serializer' => (
	isa => 'Data::Serializer',
	is => 'ro',
	default => \&_set_serializer,
	lazy => 1,
);

no Moose;

our %modules = (
	'application/json' => {
		module => 'JSON',
	},
	'application/xml' => {
		module => 'XML::Simple',
	},
	'application/yaml' => {
		module => 'YAML',
	},
	'application/x-www-form-urlencoded' => {
		module => 'FORM',
	},
);

sub _set_serializer {
	my $self = shift;
	return unless $modules{$self->type};

	my $module = $modules{$self->type}{module};
	return $module if $module eq 'FORM';

	return Data::Serializer->new(
		serializer => $module,
	);
}

sub content_type {
	my ($self) = @_;
	return $self->type;
}

sub serialize {
	my ($self, $data) = @_;
	return unless $self->serializer;

	my $result;
	try {
		$result = $self->serializer->raw_serialize($data)
	} catch {
		warn "Couldn't serialize data with " . $self->type;
	};

	return $result;
}

sub deserialize {
	my ($self, $data) = @_;
	return unless $self->serializer;

	my $result;
	try {
		$result = $self->serializer->raw_deserialize($data);
	} catch {
		warn "Couldn't deserialize data with " . $self->type;
	};

	return $result;
}

1;

__END__
=pod

=head1 NAME

Catalyst::Model::REST::Serializer

=head1 VERSION

version 0.25

=head1 AUTHOR

Kaare Rasmussen <kaare at cpan dot net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Kaare Rasmussen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

